File: assn-2/weather/broadcaster.c
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#define BUFFER_SIZE 1024
#define TERMINATE_MESSAGE "<!#TERMINATE#!>"

int main()
{
    int fd;
    char *myfifo = "/tmp/weatherMessage";
    mkfifo(myfifo, 0666);
    char buffer[BUFFER_SIZE], temp;

    printf("Starting broadcaster..........\n");
    while (1)
    {
        fd = open(myfifo, O_WRONLY);

        // Read weather info from user
        printf("Enter Weather Report: ");
        scanf("%[^\n]s", buffer);
        scanf("%c", &temp);

        // Write weather info in FIFO
        write(fd, buffer, strlen(buffer) + 1);

        // Close FIFO
        close(fd);

        printf("Terminate (Y/N): ");
        scanf("%c", &temp);
        getchar();

        if (temp == 'y' || temp == 'Y')
        {
            fd = open(myfifo, O_WRONLY);
            write(fd, TERMINATE_MESSAGE, 16);
            printf("\nTerminating broadcasting..........\n");
            break;
        }
    }
    return 0;
}

File: assn-2/weather/broadcaster.py
import os
import errno
import subprocess

TERMINATE_MESSAGE = "<!#TERMINATE#!>"


def getAllFifo():
    findFifo = subprocess.Popen(
        ["find", "./fifo", "-type", "p"], stdout=subprocess.PIPE, stderr=subprocess.PIPE
    )
    stdout, stderr = findFifo.communicate()

    if not stderr:
        stdout = str(stdout.decode("utf-8"))
        fifos = list(
            filter(
                lambda x: len(x) > 0,
                map(lambda x: x.lstrip(".").lstrip("/"), stdout.split("\n")),
            )
        )

        return fifos
    else:
        return []


def broadcastToAll(mssg):
    fifos = getAllFifo()

    if not fifos:
        if mssg != TERMINATE_MESSAGE:
            print("No listener found!")
        return

    for fifo in fifos:
        with open(fifo, "w") as fifoFile:
            fifoFile.write(mssg)


print("Starting Weather Broadcaster...")

while True:
    weatherReport = input("Enter Weather Report: ")

    fifos = getAllFifo()

    if not fifos:
        print("No Listener detected!!")
        continue

    broadcastToAll(weatherReport)

    closingConditon = input("Do you want to close broadcaster (Y/N): ")

    if closingConditon.lower() == "y":
        broadcastToAll(TERMINATE_MESSAGE)
        print("Shutting down broadcaster......")
        break